test/test_optim.py

```diff
@@ -2163,6 +2164,246 @@ def test_non_empty_state(self, device, dtype, optim_info):
+                for optim_input in optim_inputs
+                if optim_input.desc == "default"
+            ][0]
+        )
+
+        # we should receive the same output if we do the following to our models :
+        # model_default: 1.train for n_iters, 2.load state (beta1=0.0), 3.train for n_iters
+        # model_beta1_zero: 4.train for n_iters
+        n_iters = 10
+
+        def iteration(model, optim, inpt, n_iters):
+            output = None
+            for i in range(n_iters):
+                optim.zero_grad()
+                output = model(inpt)
+                loss = output.sum()
+                loss.backward(retain_graph=True)
+                optim.step()
+
+            return output
+
+        # 1. train for n_iters
+        iteration(model_default, optim_default, inpt, n_iters)
+
+        # 2. load state
+        model_default.load_state_dict(model_beta1_zero.state_dict())
+        optim_default.load_state_dict(optim_beta1_zero.state_dict())
+
+        self.assertEqual(model_beta1_zero.state_dict(), model_default.state_dict())
+        self.assertEqual(optim_beta1_zero.state_dict(), optim_default.state_dict())
+
+        inpt = torch.rand((5,), dtype=dtype, device=device)
+        # 3. train for n_iters
+        output_default = iteration(model_default, optim_default, inpt, n_iters)
+
+        # 4. train for n_iters
+        output_beta1_zero_after_load = iteration(
+            model_beta1_zero, optim_beta1_zero, inpt, n_iters
+        )
+
+        self.assertTrue(
+            torch.allclose(output_beta1_zero_after_load, output_default, atol=0.001)
+        )
+
+    @optims(
+        [optim for optim in optim_db if optim.optim_cls.__name__ == "Adam"],
+        dtypes=[torch.float32],
+    )
+    def test_correct_beta1(self, device, dtype, optim_info):
+        # we test correctness of the optimizer with beta1 = 0.0 by comparing it with the optimizer model
+        # with beta1 = 1e-6
+
+        def init_model_and_optim(optim_input):
+            model = torch.nn.Linear(5, 5)
+            torch.nn.init.ones_(model.weight)
+            torch.nn.init.ones_(model.bias)
+            model.to(dtype=dtype, device=device)
+            optim = optim_info.optim_cls(model.parameters(), **optim_input.kwargs)
+            return model, optim
+
+        def step(model, optim):
+            optim.zero_grad()
+            output = model(inpt)
+            loss = output.sum() ** 2
+            loss.backward()
+            optim.step()
+
+        inpt = torch.ones((5,), dtype=dtype, device=device)
+        optim_inputs = [
+            optim_input
+            for optim_input in optim_info.optim_inputs_func(device=device)
+            if ("betas" in optim_input.kwargs and optim_input.kwargs["betas"][0] == 0.0)
+            or optim_input.desc == "default"
+        ]
+
+        n_iters = 10
+        lr = 0.1
+
+        def run_two_models(beta1_of_default):
+            ouputs = []
+            for optim_input in optim_inputs:
```

Comment:
does this create a model per optim_input that's tested?

Reply:
If there are only 2 inputs, it may be better to not use the optim_info.optim_inputs_func(device=device) call but to just explicitly create two optimizers for clarity.

like optim_with_beta1 = ...
optim_no_beta1 = 

optim_inputs = [optim_with_beta1, optim_no_beta1]

rest of test is the same